### PR TITLE
Zarhus

### PR DESCRIPTION
**Pull request by m-iwanicki**
```sh
robot -L TRACE -b cmd_logs.txt -v rte_ip:127.0.0.1 -v config:qemu -v snipeit:no \
    -v boot_dts_from_ipxe_shell:True \
    -v dts_ipxe_link:https://boot.dasharo.com/dts/dts-rc.ipxe \
    -v features:encryption,otab \
    -v zarhus_wic_gz_file:../yocto/zarhus/zarhus-base-image-genericx86-64.rootfs.wic.gz \
    -v zarhus_swu_file:../yocto/zarhus/zarhus-swu-image-genericx86-64.rootfs.swu \
    zarhus-x86/zarhus.robot
```

Results:

```
==============================================================================
Zarhus
==============================================================================
ZHS001.206 Make sure that cukinia tests pass                          | PASS |
------------------------------------------------------------------------------
ZHS002.206 Make sure that update works :: Make sure that update vi... | PASS |
------------------------------------------------------------------------------
ZHS003.206 Make sure that failed update rollbacks to working slot ... | PASS |
------------------------------------------------------------------------------
ZHS004.206 Make sure that rollback works :: Make sure that `otab r... | PASS |
------------------------------------------------------------------------------
Zarhus                                                                | PASS |
4 tests, 4 passed, 0 failed
==============================================================================
```

[log.html](https://github.com/user-attachments/files/22901313/log.html)

On hardware:

```sh
robot -L TRACE -b cmd_logs.txt -v rte_ip:192.168.10.168 -v config:odroid-h4-plus -v snipeit:no \
    -v boot_dts_from_ipxe_shell:True \
    -v dts_ipxe_link:https://boot.dasharo.com/dts/dts-rc.ipxe \
    -v features:encryption,otab \
    -v zarhus_wic_gz_file:../yocto/zarhus/zarhus-base-image-genericx86-64.rootfs.wic.gz \
    -v target_device:nvme0n1 -t "ZHS001*" zarhus-x86/zarhus.robot
```

```
==============================================================================
Zarhus
==============================================================================
ZHS001.206 Make sure that cukinia tests pass                          | PASS |
------------------------------------------------------------------------------
Zarhus                                                                | PASS |
1 test, 1 passed, 0 failed
==============================================================================
```

[log.html](https://github.com/user-attachments/files/22908373/log.html)

---

```sh
robot -L TRACE -b command_log.txt -v rte_ip:192.168.10.168 -v config:odroid-h4-plus -v snipeit:no \
    -v boot_dts_from_ipxe_shell:True \
    -v dts_ipxe_link:https://boot.dasharo.com/dts/dts-rc.ipxe \
    -v features:encryption,otab \
    -v zarhus_bootstrap_file:../yocto/zarhus-dtrpb/bootstrap.img \
    zarhus-x86/zarhus-provisioning-preparation.robot
```

```
==============================================================================
Zarhus-Provisioning-Preparation
==============================================================================
ZPBP001.101 Make sure that network boot is disabled :: Make sure t... | PASS |
------------------------------------------------------------------------------
ZPBP002.101 Make sure that BIOS boot medium is locked :: Make sure... | PASS |
------------------------------------------------------------------------------
ZPBP003.101 Make sure that SMM BIOS write protection is enabled ::... | PASS |
------------------------------------------------------------------------------
ZPBP004.101 Make sure that Intel ME is HAP disabled :: Make sure t... | PASS |
------------------------------------------------------------------------------
ZPBP005.101 Make sure that OptionROM loading is disabled :: Make s... | PASS |
------------------------------------------------------------------------------
ZPBP006.206 Make sure that zarhus status returns expected initial ... | PASS |
------------------------------------------------------------------------------
ZPBP006.206 Make sure that Secure Boot is in Setup Mode :: Make su... | FAIL |
'Installed:      ✗ sbctl is not installed
Setup Mode:     ✗ Enabled
Secure Boot:    ✗ Disabled
Vendor Keys:    none' does not match 'Setup Mode:*Enabled'
------------------------------------------------------------------------------
Zarhus-Provisioning-Preparation                                       | FAIL |
7 tests, 6 passed, 1 failed
==============================================================================
```

After fix:

```
==============================================================================
ZPBP006.206 Make sure that zarhus status returns expected initial ... | PASS |
------------------------------------------------------------------------------
ZPBP006.206 Make sure that Secure Boot is in Setup Mode :: Make su... | PASS |
Secure Boot is in Setup Mode
------------------------------------------------------------------------------
Zarhus-Provisioning-Preparation                                       | PASS |
2 tests, 2 passed, 0 failed
==============================================================================
```

And second fix later to change `ZPBP006` -> `ZPBP007`

Couldn't test `zarhus-provisioning.robot` as Suite preparation fails in `Setup ZPB` keyword, `Should Contain    ${status}    Boot Guard: Enabled` to be exact. For some reason capsule update doesn't work.

On QEMU:

```
==============================================================================
Zarhus-Provisioning
==============================================================================
ZPB001.206 Make sure that cukinia tests pass                          | PASS |
------------------------------------------------------------------------------
Zarhus-Provisioning                                                   | PASS |
1 test, 1 passed, 0 failed
==============================================================================
```

### CODE REVIEW

File: lib/zarhus-lib.robot [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+    [Documentation]    Logs into Zarhus OS, enter decryption password if needed
+    ...
+    ...    === Requirements ===
+    ...    - Zarhus OS should be queued to boot without intervention
+    ...    - `Set Zarhus Features` keyword called beforehand e.g. in suite
+    ...    \ preparation
+    ...    - First boot setup done (set encryption password, user password changed)
+    ...
+    ...    === Effects ===
+    ...    - Power platform on
```
Comment (iwanicki92) [inline lib/zarhus-lib.robot]:
```suggestion
```

File: lib/zarhus-lib.robot [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+    Execute Command In Terminal Should Succeed
+    ...    gunzip -c /tmp/zarhus.wic.gz | dd of=/dev/${device}
+    ...    err_msg=Flashing Zarhus OS failed!
+    ...    timeout=5m
+    VAR    ${DEVICE_OS_USERNAME}=    ${os_username_old}    scope=Suite
+    VAR    ${DEVICE_OS_PASSWORD}=    ${os_password_old}    scope=Suite
+
+Install Zarhus OS
+    [Documentation]    Prepare Zarhus OS for testing. Flash it to
+    ...    ${TARGET_DEVICE}, or /dev/sda if testing on QEMU and add to
```
Comment (iwanicki92) [inline lib/zarhus-lib.robot]:
```suggestion
    ...    ${TARGET_DEVICE}, or /dev/sda if testing on QEMU and add it
    ...    as a default boot option
```

File: docs/zarhus-tests.md [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+
+#### Zarhus Provisioning Box
+
+Example command:
+
+```sh
+robot -L TRACE -b cmd_logs.txt -v rte_ip:192.168.10.168 -v config:odroid-h4-plus -v snipeit:no -v boot_dts_from_ipxe_shell:True \
+    -v dts_ipxe_link:https://boot.dasharo.com/dts/dts-rc.ipxe \
+    -v zarhus_bootstrap_file:../yocto/zarhus-dtrpb/bootstrap.img \
+    zarhus-x86/zarhus-provisioning-preparation.robot
```
Comment (iwanicki92) [inline docs/zarhus-tests.md:215]:
```suggestion
    -v features:encryption,otab \
    -v zarhus_bootstrap_file:../yocto/zarhus-dtrpb/bootstrap.img \
    zarhus-x86/zarhus-provisioning-preparation.robot
```

File: lib/zarhus-provision-lib.robot [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+    ...    - ``${disk}``: ``string`` - device on which to install Zarhus OS
+    ...    === Return Value ===
+    ...
+    ...    === Effects ===
+    ...    - Flashes ``${ZARHUS_BOOTSTRAP_FILE}`` to USB stick on hardware
+    ...    - Runs `install.sh` script from bootstrap image which installs ZPB
+    ...    \ OS on ``${disk}`` and flashes pre-configured firmware image
+    ...    - Runs first boot setup of Zarhus OS
+    [Arguments]    ${disk}=nvme0n1
+    Install ZPB OS    ${disk}
```
Comment (iwanicki92) [inline lib/zarhus-provision-lib.robot:170]:
```suggestion
    Set Zarhus Features
    Install ZPB OS    ${disk}
```

File: lib/zarhus-lib.robot [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+    ...    - Boots and configures Zarhus OS
+    Set Zarhus Features
+    Install Zarhus OS
+    Execute Reboot Command
+    Zarhus First Boot Setup
+
+Set Zarhus Features
+    [Documentation]    Return features in global ``${FEATURES}`` variable as a
+    ...    list. Features should be split with comma. Returns empty list if
+    ...    features is not defined
```
Comment (iwanicki92) [inline lib/zarhus-lib.robot]:
```suggestion
    [Documentation]    Set features from ``${FEATURES}`` variable to
    ...    ``${ZARHUS_FEATURES}`` list with Suite scope. Features should
    ...    be split with comma.
```

File: lib/zarhus-provision-lib.robot [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+    VAR    ${DEVICE_OS_USERNAME}=    ${os_username_old}    scope=Suite
+    VAR    ${DEVICE_OS_PASSWORD}=    ${os_password_old}    scope=Suite
+
+    # Make sure we can enter setup menu after flashing
+    Execute Reboot Command
+    Set DUT Response Timeout    3m
+    Enter Setup Menu Tianocore
+
+Prepare ZPB OS
+    [Documentation]    Install ZPB OS via ZPB bootstrap USB
```
Comment (iwanicki92) [inline lib/zarhus-provision-lib.robot]:
```suggestion
    [Documentation]    Install ZPB OS via ZPB bootstrap USB
    ...    and do first boot setup.
```

File: zarhus-x86/zarhus-provisioning-preparation.robot [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+
+ZPBP004.101 Make sure that Intel ME is HAP disabled
+    [Documentation]    Make sure that Intel ME is HAP disabled
+    ${state}=    Get UEFI Option    MeMode
+    Should Be Equal    ${state}    Disabled (HAP)
+
+ZPBP005.101 Make sure that OptionROM loading is disabled
+    [Documentation]    Make sure that Network Boot is disabled
+    ${state}=    Get UEFI Option    OptionROMExecutionPolicy
+    Should Be Equal    ${state}    Disable all OptionROMs
```
Comment (iwanicki92) [inline zarhus-x86/zarhus-provisioning-preparation.robot]:
```suggestion
    Should Contain    ${state}    Disable all OptionROMs
```

File: zarhus-x86/zarhus.robot [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+    # First one is printed during reboot/shutdown when entering initramfs
+    Read From Terminal Until    Starting systemd-udevd
+    Read From Terminal Until    Starting systemd-udevd
+    # Power On should reboot before we boot into OS and confirm that update worked
+    Boot Zarhus OS
+    ${current_part2}=    Execute Command In Terminal    fw_printenv otab_part_current
+    Should Be Equal    ${current_part}    ${current_part2}
+
+ZHS004.206 Make sure that rollback works
+    [Documentation]    Make sure that `otab rollback` works
```
Comment (iwanicki92) [inline zarhus-x86/zarhus.robot]:
```suggestion
    [Documentation]    Make sure that `otab rollback` will result in 
    ...    booting into inactive slot after reboot.
```

File: os-config/206-credentials.py [hunk 9]
```diff
@@ -0,0 +1,10 @@
+# SPDX-FileCopyrightText: 2025 3mdeb <contact@3mdeb.com>
```
Comment (iwanicki92) [inline os-config/206-credentials.py:1]: Missing changes to `environment-test-ids.py`, probably lost during rebase.

File: lib/usb-hid-msc-lib.robot [hunk 10]
```diff
@@ -74,3 +75,31 @@ Check USB Stick Detection In Edk2
...
+    ...
+    ...    === Effects ===
+    ${devices}=    Execute Command In Terminal
+    ...    ls -A1 /dev/disk/by-id/"usb-"* | xargs readlink -f | sed 's|/dev/||g'
+    @{dev_list}=    Split To Lines    ${devices}
+    FOR    ${device}    IN    @{dev_list}
+        # Removable USB storage (size > 0), that is not partition
+        VAR    ${command}=
+        ...    test ! -f "/sys/class/block/${device}/partition"
+        ...    test "\$(cat /sys/class/block/${device}/size)"
```
Comment (iwanicki92) [inline lib/usb-hid-msc-lib.robot]:
```suggestion
        ...    test "\$(cat /sys/class/block/${device}/size)" -gt 0
```

File: zarhus-x86/zarhus-provisioning-preparation.robot [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+Library             SSHLibrary    timeout=90 seconds
+Library             RequestsLibrary
+Resource            ../keywords.robot
+Resource            ../keys.robot
+Resource            ../variables.robot
+Resource            ../lib/zarhus-provision-lib.robot
+
+Suite Setup         Run Keywords
+...                     Prepare Test Suite
+...                     AND
```
Comment (iwanicki92) [inline zarhus-x86/zarhus-provisioning-preparation.robot:17]:
```suggestion
...                     Prepare Test Suite
...                     AND
...                     Skip If    '${MANUFACTURER}' == 'QEMU'
...                     AND
```

File: lib/zarhus-lib.robot [hunk 12]
```diff
@@ -0,0 +1,208 @@
...
+    Enter Shell In DTS
+    Execute Command In Terminal    systemctl start sshd
+    Send File To DUT    ${zarhus_os_file}    /tmp/zarhus.wic.gz
+    Execute Command In Terminal Should Succeed
+    ...    test -b /dev/${device}    err_msg=/dev/${device} block device doesn't exist!
+    Execute Command In Terminal Should Succeed
+    ...    gunzip -c /tmp/zarhus.wic.gz | dd of=/dev/${device}
+    ...    err_msg=Flashing Zarhus OS failed!
+    ...    timeout=5m
+    VAR    ${DEVICE_OS_USERNAME}=    ${os_username_old}    scope=Suite
```
Comment (iwanicki92) [inline lib/zarhus-lib.robot:88]:
```suggestion
    # Using `scope=Suite` because `scope=Test` resulted in those variables
    # having root/${EMPTY} values in tests (after suite setup finished)
    VAR    ${DEVICE_OS_USERNAME}=    ${os_username_old}    scope=Suite
```

File: lib/QemuMonitor.py [hunk 13]
```diff
@@ -181,6 +156,7 @@ def usb_add(self, img_path, name="unnamed"):
             "driver": "usb-storage",
             "id": self._usb_devid(name),  # "usbdisk",
             "drive": self._usb_nodename(name),
+            "removable": "on" if removable else "off",
```
Comment (iwanicki92) [inline lib/QemuMonitor.py]:
```suggestion
            "removable": removable,
```

File: docs/zarhus-tests.md [hunk 14]
```diff
@@ -0,0 +1,223 @@
...
+    [Test Variables](#test-variables)
+
+### Hardware
+
+#### Zarhus OS
+
+Example test run that succeeded:
+
+```sh
+robot -L TRACE -b cmd_logs.txt -v rte_ip:192.168.10.168 -v config:odroid-h4-plus -v snipeit:no \
```
Comment (iwanicki92) [inline docs/zarhus-tests.md]:
```suggestion
robot -L TRACE -b command_log.txt -v rte_ip:192.168.10.168 -v config:odroid-h4-plus -v snipeit:no \
```

File: docs/zarhus-tests.md [hunk 15]
```diff
@@ -0,0 +1,223 @@
...
+    -v zarhus_wic_gz_file:../zarhus-base-image-genericx86-64.rootfs.wic.gz \
+    -v target_device:nvme0n1 -t "ZHS001*" zarhus-x86/zarhus.robot
+```
+
+#### Zarhus Provisioning Box
+
+Example command:
+
+```sh
+robot -L TRACE -b cmd_logs.txt -v rte_ip:192.168.10.168 -v config:odroid-h4-plus -v snipeit:no -v boot_dts_from_ipxe_shell:True \
```
Comment (iwanicki92) [inline docs/zarhus-tests.md]:
```suggestion
robot -L TRACE -b command_log.txt -v rte_ip:192.168.10.168 -v config:odroid-h4-plus -v snipeit:no -v boot_dts_from_ipxe_shell:True \
```

File: lib/zarhus-provision-lib.robot [hunk 16]
```diff
@@ -0,0 +1,121 @@
...
+    ...    - Runs `install.sh` script from bootstrap image which installs ZPB
+    ...    \ OS on ``${disk}`` and flashes pre-configured firmware image
+    ...    - Runs first boot setup of Zarhus OS
+    [Arguments]    ${disk}=nvme0n1
+    Install ZPB OS    ${disk}
+    Boot Dasharo Tools Suite    iPXE
+    Enter Shell In DTS
+    # Make sure Zarhus is first boot entry
+    Execute Command In Terminal Should Succeed
+    ...    efibootmgr --disk "/dev/${disk}" --part 1 --create --label "ZarhusOS" --loader '\\EFI\\BOOT\\bootx64.efi'
```
Comment (iwanicki92) [inline lib/zarhus-provision-lib.robot:181]:
```suggestion
    Execute Command In Terminal Should Succeed
    ...    mount -o remount,rw efivarfs
    ...    Failed to remount efivarfs as read-writable
    Execute Command In Terminal Should Succeed
    ...    efibootmgr --disk "/dev/${disk}" --part 1 --create --label "ZarhusOS" --loader '\\EFI\\BOOT\\bootx64.efi'
```

File: lib/zarhus-lib.robot [hunk 17]
```diff
@@ -0,0 +1,211 @@
...
+    VAR    ${os_password_old}=    ${DEVICE_OS_PASSWORD}
+    VAR    ${DEVICE_OS_USERNAME}=    root    scope=Test
+    VAR    ${DEVICE_OS_PASSWORD}=    ${EMPTY}    scope=Test
+    Enter Shell In DTS
+    Execute Command In Terminal    systemctl start sshd
+    Send File To DUT    ${zarhus_os_file}    /tmp/zarhus.wic.gz
+    Execute Command In Terminal Should Succeed
+    ...    test -b /dev/${device}    err_msg=/dev/${device} block device doesn't exist!
+    Execute Command In Terminal Should Succeed
+    ...    gunzip -c /tmp/zarhus.wic.gz | dd of=/dev/${device}
```
Comment (iwanicki92) [inline lib/zarhus-lib.robot]: `set -o pipefail` in case `gunzip` fails

Comment (m-iwanicki) [reply lib/zarhus-lib.robot]: [c02c9...ee3d0](https://github.com/Dasharo/open-source-firmware-validation/compare/c02c919707a6257602e7e9ba6273e0b996981485..ee3d0b4b67b711e74c5c889cedf337ea61c7f941)

File: docs/zarhus-tests.md [hunk 18]
```diff
@@ -0,0 +1,242 @@
...
+genericx86-64:~$ lsblk | grep 300M
+sdb                  8:16   1  300M  0 disk
+sdc                  8:32   1  300M  0 disk
+genericx86-64:~$ cat /sys/class/block/sdb/removable
+1
+```
+
+## Automatic tests
+
+**Important**: To run OSFV tests on Zarhus OS with encryption feature enabled
```
Comment (ghost) [inline docs/zarhus-tests.md]: This should highlight this should be done at building stage. We're mentioning provisioning box here anyway so I think we might add the path (not link) for which file to edit as well. Spares the reproducer time to figure it out.

File: docs/zarhus-tests.md [hunk 19]
```diff
@@ -0,0 +1,242 @@
...
+    -v boot_dts_from_ipxe_shell:True \
+    -v dts_ipxe_link:https://boot.dasharo.com/dts/dts-rc.ipxe \
+    -v features:encryption,otab \
+    -v zarhus_wic_gz_file:../zarhus-base-image-genericx86-64.rootfs.wic.gz \
+    -v target_device:nvme0n1 -t "ZHS001*" zarhus-x86/zarhus.robot
+```
+
+#### Zarhus Provisioning Box
+
+Before starting, flash firmware manually.
```
Comment (ghost) [inline docs/zarhus-tests.md]: Please add short explanation why this step needs to be done, there is no context to it. Might as well add link to docs dasharo firmares for convenience, currently odroid is supported anyway.

File: lib/usb-hid-msc-lib.robot [hunk 20]
```diff
@@ -74,3 +75,31 @@ Check USB Stick Detection In Edk2
...
+    ...
+    ...    === Effects ===
+    ${devices}=    Execute Command In Terminal
+    ...    ls -A1 /dev/disk/by-id/"usb-"* | xargs readlink -f | sed 's|/dev/||g'
+    @{dev_list}=    Split To Lines    ${devices}
+    FOR    ${device}    IN    @{dev_list}
+        # Removable USB storage (size > 0), that is not partition
+        VAR    ${command}=
+        ...    test ! -f "/sys/class/block/${device}/partition"
+        ...    test "\$(cat /sys/class/block/${device}/size)" -gt 0
```
Comment (ghost) [inline lib/usb-hid-msc-lib.robot:99]: Very corner case I suppose but specifying size might spare somebody debug in the future.

**Review by ghost — COMMENTED** [review]
Contentwise looks good to me, small nitpicks for now. I'll proceed to running tests.

File: lib/usb-hid-msc-lib.robot (hunk 20 above)
Comment (m-iwanicki) [reply lib/usb-hid-msc-lib.robot:99]: I'm not sure I understand? I verify size (`-gt 0`) is greater than 0

File: docs/zarhus-tests.md [hunk 21]
```diff
@@ -0,0 +1,242 @@
...
+
+Example command to run tests:
+
+```sh
```
Comment (ghost) [inline docs/zarhus-tests.md:210]: It must be highlighted how to prepare that bootsrap file and how to mount it and edit if needed.

**Comment by ghost** [comment]
Tested provisioning box on hardware, details below

<details>
<summary> HW PROV.BOX </summary>

All tests have passed beside one

```log
(venv) mkusiak in ~/Development/dts/open-source-firmware-validation on zarhus ● λ robot -L TRACE -b command_log.txt -v rte_ip:192.168.10.168 -v config:odroid-h4-plus -v snipeit:no -v boot_dts_from_ipxe_shell:True \
    -v dts_ipxe_link:https://boot.dasharo.com/dts/dts-rc.ipxe \
    -v features:encryption,otab \
    -v zarhus_bootstrap_file:"$bootstrap" \
    zarhus-x86/zarhus-provisioning-preparation.robot
==============================================================================
Zarhus-Provisioning-Preparation                                               
==============================================================================
ZPBP001.101 Make sure that network boot is disabled :: Make sure t... | PASS |
------------------------------------------------------------------------------
ZPBP002.101 Make sure that BIOS boot medium is locked :: Make sure... | PASS |
------------------------------------------------------------------------------
ZPBP003.101 Make sure that SMM BIOS write protection is enabled ::... | PASS |
------------------------------------------------------------------------------
ZPBP004.101 Make sure that Intel ME is HAP disabled :: Make sure t... | PASS |
------------------------------------------------------------------------------
ZPBP005.101 Make sure that OptionROM loading is disabled :: Make s... | PASS |
------------------------------------------------------------------------------
ZPBP006.206 Make sure that zarhus status returns expected initial ... | FAIL |
low != high
------------------------------------------------------------------------------
ZPBP007.206 Make sure that Secure Boot is in Setup Mode :: Make su... | PASS |
Secure Boot is in Setup Mode
------------------------------------------------------------------------------
Zarhus-Provisioning-Preparation                                       | FAIL |
7 tests, 6 passed, 1 failed
==============================================================================
Debug:   /home/mkusiak/Development/dts/open-source-firmware-validation/command_log.txt
Output:  /home/mkusiak/Development/dts/open-source-firmware-validation/output.xml
Log:     /home/mkusiak/Development/dts/open-source-firmware-validation/log.html
Report:  /home/mkusiak/Development/dts/open-source-firmware-validation/report.html
```

One of the tests failed due to improper state of power LED, this is a known issue.

image

</details>

Will proceed to test zarhus os.

File: docs/zarhus-tests.md (hunk 19 above)
Comment (m-iwanicki) [reply docs/zarhus-tests.md]: > currently

and possibly not always [21147..82184](https://github.com/Dasharo/open-source-firmware-validation/compare/21147742280aa7e4fe02407100e09f4986bb3f90..821849f25a1e6248235c70942eb7516f4f323a18)

File: docs/zarhus-tests.md (hunk 21 above)
Comment (m-iwanicki) [reply docs/zarhus-tests.md:210]: [21147..82184](https://github.com/Dasharo/open-source-firmware-validation/compare/21147742280aa7e4fe02407100e09f4986bb3f90..821849f25a1e6248235c70942eb7516f4f323a18)

File: docs/zarhus-tests.md (hunk 18 above)
Comment (m-iwanicki) [reply docs/zarhus-tests.md]: [21147..82184](https://github.com/Dasharo/open-source-firmware-validation/compare/21147742280aa7e4fe02407100e09f4986bb3f90..821849f25a1e6248235c70942eb7516f4f323a18)

File: docs/zarhus-tests.md [hunk 22]
```diff
@@ -0,0 +1,260 @@
...
+genericx86-64:~$ cat /sys/class/block/sdb/removable
+1
+```
+
+## Automatic tests
+
+**Important**: To run OSFV tests on Zarhus OS with encryption feature enabled
+you have to make sure that kernel command line contains `console=ttyS0,115200`
+instead of `console=tty1`, as disk encryption password prompt displayed in
+initramfs shows only on that `tty`. Running tests with wrong image will result
```
Comment (ghost) [inline docs/zarhus-tests.md]: This is a bit misleading, as it can be understood as "if encryption layer is added it is necessary to edit CMDLINE".

File: docs/zarhus-tests.md [hunk 23]
```diff
@@ -0,0 +1,260 @@
...
+    [Test Variables](#test-variables)
+
+### Hardware
+
+#### Zarhus OS
+
+Example test run that succeeded:
+
+```sh
+robot -L TRACE -b command_log.txt -v rte_ip:192.168.10.168 -v config:odroid-h4-plus -v snipeit:no \
```
Comment (ghost) [inline docs/zarhus-tests.md:193]: Might be wise to list what layers shall be built in. Eg. this won't run without encryption support added.

**Comment by ghost** [comment]
Hardware ZarhusOS tests pass

<details><summary>ZarhusOS HW</summary>
<p>

```log
==============================================================================
Zarhus                                                                        
==============================================================================
ZHS001.206 Make sure that cukinia tests pass                          | PASS |
------------------------------------------------------------------------------
Zarhus                                                                | PASS |
1 test, 1 passed, 0 failed
==============================================================================
Debug:   /home/mkusiak/Development/dts/open-source-firmware-validation/command_log.txt
Output:  /home/mkusiak/Development/dts/open-source-firmware-validation/output.xml
Log:     /home/mkusiak/Development/dts/open-source-firmware-validation/log.html
Report:  /home/mkusiak/Development/dts/open-source-firmware-validation/report.html
```

</p>
</details> 

Will try qemu next

File: docs/zarhus-tests.md [hunk 24]
```diff
@@ -0,0 +1,260 @@
...
+    -v boot_dts_from_ipxe_shell:True \
+    -v dts_ipxe_link:https://boot.dasharo.com/dts/dts-rc.ipxe \
+    -v features:encryption,otab \
+    -v zarhus_wic_gz_file:../zarhus-base-image-genericx86-64.rootfs.wic.gz \
+    -v target_device:nvme0n1 -t "ZHS001*" zarhus-x86/zarhus.robot
```
Comment (ghost) [inline docs/zarhus-tests.md]: Let's run all tests by default.

**Comment by ghost** [comment]
Rerun full test set this time, all tests do pass. 

<details><summary>ZarhusOS full scope HW</summary>
<p>

```log
==============================================================================
Zarhus                                                                        
==============================================================================
ZHS001.206 Make sure that cukinia tests pass                          | PASS |
------------------------------------------------------------------------------
ZHS002.206 Make sure that update works :: Make sure that update vi... | PASS |
------------------------------------------------------------------------------
ZHS003.206 Make sure that failed update rollbacks to working slot ... | PASS |
------------------------------------------------------------------------------
ZHS004.206 Make sure that rollback works :: Make sure that `otab r... | PASS |
------------------------------------------------------------------------------
Zarhus                                                                | PASS |
4 tests, 4 passed, 0 failed
==============================================================================
Debug:   /home/mkusiak/Development/dts/open-source-firmware-validation/command_log.
Output:  /home/mkusiak/Development/dts/open-source-firmware-validation/output.xml
Log:     /home/mkusiak/Development/dts/open-source-firmware-validation/log.html
Report:  /home/mkusiak/Development/dts/open-source-firmware-validation/report.html
```

</p>
</details> 

Skipping doing tests on QEMU, waste of time if hardware works.

**Review by ghost — CHANGES_REQUESTED** [review]

File: docs/zarhus-tests.md (hunk 23 above)
Comment (m-iwanicki) [reply docs/zarhus-tests.md:193]: > Might be wise to list what layers shall be built in

The ones you want to test. I mentioned in docs that only `meta-zarhus-hardening` is required.
Tests should be skipped if they are not supported e.g.: https://github.com/Dasharo/open-source-firmware-validation/blob/308dc4a10fdcc7e2973cda23b50c6a3d784a4261/dasharo-compatibility/audio-subsystem.robot#L54-L56

Comment (m-iwanicki) [reply docs/zarhus-tests.md:193]: > Eg. this won't run without encryption support added.

I'd assume that person testing knows what's he's doing. There is variable `-v features:encryption,otab` which lists features your image supports, described more extensively in `### Test Variables`. Expanded a little to mention which `.yml` files are required for this feature.

[82184..d10cc](https://github.com/Dasharo/open-source-firmware-validation/compare/821849f25a1e6248235c70942eb7516f4f323a18..d10cc6344d54e1cd8c1e759af00b9206e860632b)

File: docs/zarhus-tests.md (hunk 24 above)
Comment (m-iwanicki) [reply docs/zarhus-tests.md]: [82184..d10cc](https://github.com/Dasharo/open-source-firmware-validation/compare/821849f25a1e6248235c70942eb7516f4f323a18..d10cc6344d54e1cd8c1e759af00b9206e860632b)

File: docs/zarhus-tests.md (hunk 22 above)
Comment (m-iwanicki) [reply docs/zarhus-tests.md]: [82184..d10cc](https://github.com/Dasharo/open-source-firmware-validation/compare/821849f25a1e6248235c70942eb7516f4f323a18..d10cc6344d54e1cd8c1e759af00b9206e860632b)

File: docs/zarhus-tests.md (hunk 23 above)
Comment (m-iwanicki) [reply docs/zarhus-tests.md:193]: Also added skip to update tests if no otab feature available: [d10cc..39dae](https://github.com/Dasharo/open-source-firmware-validation/compare/d10cc6344d54e1cd8c1e759af00b9206e860632b..39daed090aef152e0e6547911eb07d0ae048402b):

```
robot -L TRACE -b cmd_logs.txt -v rte_ip:192.168.10.168 -v config:odroid-h4-plus -v snipeit:no -v boot_dts_from_ipxe_shell:True \
    -v dts_ipxe_link:https://boot.dasharo.com/dts/dts-rc.ipxe \
        -v features:encryption \
        -v zarhus_wic_gz_file:../yocto/zarhus/zarhus-base-image-genericx86-64.rootfs.wic.gz \
        -v target_device:nvme0n1 -t "ZHS00[234]*" zarhus-x86/zarhus.robot
==============================================================================
Zarhus
==============================================================================
ZHS002.206 Make sure that update works :: Make sure that update vi... | SKIP |
ZHS002.206 not supported
------------------------------------------------------------------------------
ZHS003.206 Make sure that failed update rollbacks to working slot ... | SKIP |
ZHS003.206 not supported
------------------------------------------------------------------------------
ZHS004.206 Make sure that rollback works :: Make sure that `otab r... | SKIP |
ZHS004.206 not supported
------------------------------------------------------------------------------
Zarhus                                                                | SKIP |
3 tests, 0 passed, 0 failed, 3 skipped
==============================================================================
```

**Review by ghost — APPROVED** [review]
LGTM

**Comment by ghost** [comment]
Rebased on top of develop. Waiting for checks to complete.
